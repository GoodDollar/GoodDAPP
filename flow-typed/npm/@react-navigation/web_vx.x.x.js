// flow-typed signature: 6afd15d96d524b48dd70fa758373d370
// flow-typed version: <<STUB>>/@react-navigation/web_v^1.0.0-alpha.7/flow_v0.89.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@react-navigation/web'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@react-navigation/web' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@react-navigation/web/dist/createBrowserApp' {
  declare module.exports: any;
}

declare module '@react-navigation/web/dist/handleServerRequest' {
  declare module.exports: any;
}

declare module '@react-navigation/web/dist/index' {
  declare module.exports: any;
}

declare module '@react-navigation/web/dist/Link' {
  declare module.exports: any;
}

declare module '@react-navigation/web/src/__tests__/createBrowserApp-test' {
  declare module.exports: any;
}

declare module '@react-navigation/web/src/createBrowserApp' {
  declare module.exports: any;
}

declare module '@react-navigation/web/src/handleServerRequest' {
  declare module.exports: any;
}

declare module '@react-navigation/web/src/index' {
  declare module.exports: any;
}

declare module '@react-navigation/web/src/Link' {
  declare module.exports: any;
}

// Filename aliases
declare module '@react-navigation/web/dist/createBrowserApp.js' {
  declare module.exports: $Exports<'@react-navigation/web/dist/createBrowserApp'>;
}
declare module '@react-navigation/web/dist/handleServerRequest.js' {
  declare module.exports: $Exports<'@react-navigation/web/dist/handleServerRequest'>;
}
declare module '@react-navigation/web/dist/index.js' {
  declare module.exports: $Exports<'@react-navigation/web/dist/index'>;
}
declare module '@react-navigation/web/dist/Link.js' {
  declare module.exports: $Exports<'@react-navigation/web/dist/Link'>;
}
declare module '@react-navigation/web/src/__tests__/createBrowserApp-test.js' {
  declare module.exports: $Exports<'@react-navigation/web/src/__tests__/createBrowserApp-test'>;
}
declare module '@react-navigation/web/src/createBrowserApp.js' {
  declare module.exports: $Exports<'@react-navigation/web/src/createBrowserApp'>;
}
declare module '@react-navigation/web/src/handleServerRequest.js' {
  declare module.exports: $Exports<'@react-navigation/web/src/handleServerRequest'>;
}
declare module '@react-navigation/web/src/index.js' {
  declare module.exports: $Exports<'@react-navigation/web/src/index'>;
}
declare module '@react-navigation/web/src/Link.js' {
  declare module.exports: $Exports<'@react-navigation/web/src/Link'>;
}
