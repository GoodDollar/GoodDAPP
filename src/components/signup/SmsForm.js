// @flow
import React, { useEffect, useState } from 'react'
import { KeyboardAvoidingView } from 'react-native'
import { isIOS } from '../../lib/utils/platform'
import logger from '../../lib/logger/pino-logger'
import API, { getErrorMessage } from '../../lib/API/api'
import { getDesignRelativeHeight } from '../../lib/utils/sizes'
import { withStyles } from '../../lib/styles'
import SpinnerCheckMark from '../common/animations/SpinnerCheckMark'
import Section from '../common/layout/Section'
import ErrorText from '../common/form/ErrorText'
import OtpInput from '../common/form/OtpInput'
import CustomWrapper from './signUpWrapper'
import type { SignupState } from './SignupState'

const log = logger.child({ from: 'SmsForm' })

type Props = {
  phone: string,
  data: SignupState,
  doneCallback: ({ isPhoneVerified: boolean }) => null,
  screenProps: any,
}

export type SMSRecord = {
  smsValidated: boolean,
  sentSMS?: boolean,
}

type State = SMSRecord & {
  errorMessage: string,
  sendingCode: boolean,
  renderButton: boolean,
  resentCode: boolean,
  loading: boolean,
  otp: Array<string>,
}

const NumInputs: number = 6

class SmsForm extends React.Component<Props, State> {
  state = {
    smsValidated: false,
    sentSMS: false,
    errorMessage: '',
    sendingCode: false,
    showWait: true,
    resentCode: false,
    loading: false,
    otp: Array(NumInputs).fill(null),
  }

  handleChange = async (otp: array) => {
    const otpValue = otp.filter(val => val).join('')
    if (otpValue.replace(/ /g, '').length === NumInputs) {
      this.setState({
        loading: true,
        otp,
      })
      try {
        await this.verifyOTP(otpValue)
        this.handleSubmit()
      } catch (e) {
        if (e.ok === 0) {
          log.warn('Verify otp failed', e.message, e)
        } else {
          log.error('Verify otp failed', e.message, e)
        }

        this.setState({
          errorMessage: e.message || e,
          loading: false,
        })
      }
    } else {
      this.setState({
        errorMessage: '',
        otp,
      })
    }
  }

  handleSubmit = async () => {
    this.setState({ loading: false })
    await this.props.screenProps.doneCallback({ smsValidated: true })
  }

  // eslint-disable-next-line class-methods-use-this
  verifyOTP(otp: string) {
    return API.verifyMobile({ otp })
  }

  handleRetryWithCall = async () => {
    await this.handleRetry({ otpChannel: 'call' })
  }

  handleRetry = async (options: any) => {
    const { otpChannel } = options
    this.setState({ sendingCode: true, otp: Array(NumInputs).fill(null), errorMessage: '' })
    let { retryFunctionName } = this.props.screenProps

    retryFunctionName = retryFunctionName || 'sendOTP'

    try {
      await API[retryFunctionName]({ ...this.props.screenProps.data, otpChannel })
      this.setState({ sendingCode: false, resentCode: true })
    } catch (e) {
      const errorMessage = getErrorMessage(e)
      const exception = new Error(errorMessage)

      log.error('Resend sms code failed', errorMessage, exception)

      this.setState({
        errorMessage,
        sendingCode: false,
        resentCode: false,
      })
    }
  }

  render() {
    const { errorMessage, otp, renderButton, sendingCode, resentCode } = this.state
    const { styles } = this.props

    return (
      <KeyboardAvoidingView behavior={isIOS ? 'padding' : 'height'} style={styles.keyboardAvoidWrapper}>
        <CustomWrapper handleSubmit={this.handleSubmit} footerComponent={() => <React.Fragment />}>
          <Section grow justifyContent="flex-start">
            <Section.Stack justifyContent="flex-start" style={styles.container}>
              <Section.Row justifyContent="center">
                <Section.Title color="darkGray" fontSize={22} fontWeight="500" textTransform="none">
                  {'Enter the verification code\nsent to your phone'}
                </Section.Title>
              </Section.Row>
              <Section.Stack justifyContent="center" style={styles.bottomContent}>
                <OtpInput
                  shouldAutoFocus
                  numInputs={NumInputs}
                  onChange={this.handleChange}
                  hasErrored={errorMessage !== ''}
                  errorStyle={styles.errorStyle}
                  value={otp}
                  placeholder="*"
                  isInputNum={true}
                  aside={[3]}
                />
                <ErrorText error={errorMessage} />
              </Section.Stack>
            </Section.Stack>
            <Section.Row alignItems="center" justifyContent="center" style={styles.row}>
              <SMSAction
                styles={styles}
                sendingCode={sendingCode}
                resentCode={resentCode}
                renderButton={renderButton}
                handleRetry={this.handleRetry}
                handleRetryWithCall={this.handleRetryWithCall}
                onFinish={() => {
                  //reset smsaction state
                  this.setState({ resentCode: false })
                }}
              />
            </Section.Row>
          </Section>
        </CustomWrapper>
      </KeyboardAvoidingView>
    )
  }
}

const SMSAction = ({ handleRetry, handleRetryWithCall, resentCode, sendingCode, onFinish, styles }) => {
  const [showWait, setWait] = useState(true)

  useEffect(() => {
    if (showWait) {
      setTimeout(() => {
        setWait(false)
      }, 10000)
    }
  }, [showWait])

  if (showWait === false) {
    return (
      <SpinnerCheckMark
        height={100}
        loading={sendingCode}
        success={resentCode}
        onFinish={() => {
          setWait(true)
          onFinish()
        }}
      >
        <Section.Text
          textDecorationLine="underline"
          fontWeight="medium"
          fontSize={14}
          color="primary"
          onPress={handleRetry}
          style={styles.sendCodeAgainButton}
        >
          Send the code again
        </Section.Text>
        <Section.Text
          textDecorationLine="underline"
          fontWeight="medium"
          fontSize={14}
          color="primary"
          onPress={handleRetryWithCall}
        >
          Send via voice call
        </Section.Text>
      </SpinnerCheckMark>
    )
  }

  return (
    <Section.Text fontSize={14} color="gray80Percent">
      Please wait a few seconds until the SMS arrives
    </Section.Text>
  )
}

const getStylesFromProps = ({ theme }) => ({
  keyboardAvoidWrapper: {
    width: '100%',
  },
  informativeParagraph: {
    margin: '1em',
  },
  buttonWrapper: {
    alignContent: 'stretch',
    flexDirection: 'column',
    display: 'flex',
    justifyContent: 'space-between',
  },
  button: {
    justifyContent: 'center',
    width: '100%',
    height: 60,
  },
  row: {
    marginVertical: theme.sizes.defaultDouble,
  },
  errorStyle: {
    borderBottomWidth: 1,
    borderBottomColor: theme.colors.red,
    color: theme.colors.red,
  },
  container: {
    minHeight: getDesignRelativeHeight(200),
    height: getDesignRelativeHeight(200),
  },
  bottomContent: {
    marginTop: 'auto',
    marginBottom: theme.sizes.defaultDouble,
  },
  sendCodeAgainButton: {
    marginRight: 'auto',
  },
})

export default withStyles(getStylesFromProps)(SmsForm)
