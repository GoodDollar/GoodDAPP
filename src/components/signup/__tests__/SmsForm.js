import { createSwitchNavigator } from '@react-navigation/core'
import { createBrowserApp } from '@react-navigation/web'
import React from 'react'
import renderer from 'react-test-renderer'
import { StoresWrapper, withThemeProvider } from '../../../__tests__/__util__'

// Note: test renderer must be required after react-native.

import SmsForm from '../SmsForm'

describe('SmsForm', () => {
  it('matches snapshot', async () => {
    const WebRouter = withThemeProvider(createBrowserApp(createSwitchNavigator({ SMS: SmsForm })))
    let component

    await renderer.act(
      // eslint-disable-next-line require-await
      async () =>
        (component = renderer.create(
          <StoresWrapper>
            <WebRouter />
          </StoresWrapper>,
        )),
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
