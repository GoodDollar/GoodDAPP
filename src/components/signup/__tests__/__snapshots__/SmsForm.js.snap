// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmsForm matches snapshot 1`] = `
Array [
  <div
    className="css-view-1dbjc4n r-flex-13awgt0 r-pointerEvents-12vffkv"
  >
    <div
      className="css-view-1dbjc4n r-width-13qz1uu"
    >
      <div
        className="css-view-1dbjc4n r-flex-13awgt0 r-paddingBottom-97e31f r-paddingLeft-bv2aro r-paddingRight-hxflta r-paddingTop-10xqauy"
      >
        <div
          className="css-view-1dbjc4n r-display-6koalj r-flexDirection-eqz5dr r-flexGrow-16y2uox r-maxHeight-87ditf r-overflowX-lltvgl r-overflowY-1rnoaur r-padding-nsbfu8 r-position-bnwqim r-width-13qz1uu"
          data-name="viewWrapper"
          style={
            Object {
              "backgroundColor": "rgba(0,0,0,0.00)",
            }
          }
        >
          <div
            className="css-view-1dbjc4n r-backgroundColor-1niwhzg r-borderRadius-1f0042m r-flexGrow-16y2uox r-padding-t60dpp r-paddingHorizontal-1e081e0 r-paddingVertical-1yzf0co"
          >
            <div
              className="css-view-1dbjc4n r-alignItems-1oszu61 r-flexDirection-eqz5dr r-flexGrow-16y2uox"
              style={
                Object {
                  "WebkitBoxPack": "space-evenly",
                  "WebkitJustifyContent": "space-evenly",
                  "justifyContent": "space-evenly",
                  "msFlexPack": "space-evenly",
                }
              }
            >
              <div
                className="css-view-1dbjc4n r-backgroundColor-14lw9ot r-borderRadius-1f0042m r-flexGrow-16y2uox r-paddingHorizontal-1e081e0 r-paddingVertical-1yzf0co"
                style={
                  Object {
                    "WebkitBoxPack": "start",
                    "WebkitJustifyContent": "flex-start",
                    "justifyContent": "flex-start",
                    "msFlexPack": "start",
                  }
                }
              >
                <div
                  className="css-view-1dbjc4n r-alignItems-1oszu61 r-flexDirection-eqz5dr r-height-1w2pmg r-minHeight-ifefl9"
                  style={
                    Object {
                      "WebkitBoxPack": "start",
                      "WebkitJustifyContent": "flex-start",
                      "justifyContent": "flex-start",
                      "msFlexPack": "start",
                    }
                  }
                >
                  <div
                    className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz"
                    style={
                      Object {
                        "WebkitBoxPack": "center",
                        "WebkitJustifyContent": "center",
                        "justifyContent": "center",
                        "msFlexPack": "center",
                      }
                    }
                  >
                    <div
                      className="css-text-901oao r-color-cddv6f r-fontFamily-vnw8o6 r-fontSize-hiw28u r-fontWeight-majxgm r-lineHeight-qvk6io r-marginVertical-c8eef1 r-textAlign-q4m81j r-textDecorationLine-13wfysu r-textTransform-3twk1y"
                      dir="auto"
                    >
                      Enter the verification code
sent to your phone
                    </div>
                  </div>
                  <div
                    className="css-view-1dbjc4n r-alignItems-1oszu61 r-flexDirection-eqz5dr r-marginBottom-1ifxtd0 r-marginTop-1t01tom"
                    style={
                      Object {
                        "WebkitBoxPack": "center",
                        "WebkitJustifyContent": "center",
                        "justifyContent": "center",
                        "msFlexPack": "center",
                      }
                    }
                  >
                    <div
                      className="css-view-1dbjc4n r-display-6koalj r-flexDirection-18u37iz r-justifyContent-1d7fvdj r-marginBottom-5oul0u"
                    >
                      <div
                        className="css-view-1dbjc4n r-alignItems-1awozwy r-display-6koalj r-flex-13awgt0 r-paddingHorizontal-mk0yit"
                      >
                        <input
                          autoCapitalize="sentences"
                          autoComplete="on"
                          autoCorrect="on"
                          className="css-textinput-11aywtz r-borderBottomColor-1uy9gz5 r-borderBottomWidth-wgabs5 r-borderLeftWidth-gxnn5r r-borderRightWidth-fnigne r-borderTopWidth-13yce4e r-fontSize-hiw28u r-marginBottom-p1pxzi r-marginVertical-r0h9e2 r-paddingVertical-rjfia r-textAlign-q4m81j r-width-13qz1uu"
                          dir="auto"
                          disabled={false}
                          enterKeyHint="next"
                          maxLength={1}
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          onKeyDown={[Function]}
                          onSelect={[Function]}
                          placeholder="*"
                          readOnly={false}
                          spellCheck={true}
                          type="tel"
                          value=""
                        />
                      </div>
                      <div
                        className="css-view-1dbjc4n r-alignItems-1awozwy r-display-6koalj r-flex-13awgt0 r-paddingHorizontal-mk0yit"
                      >
                        <input
                          autoCapitalize="sentences"
                          autoComplete="on"
                          autoCorrect="on"
                          className="css-textinput-11aywtz r-borderBottomColor-1uy9gz5 r-borderBottomWidth-wgabs5 r-borderLeftWidth-gxnn5r r-borderRightWidth-fnigne r-borderTopWidth-13yce4e r-fontSize-hiw28u r-marginBottom-p1pxzi r-marginVertical-r0h9e2 r-paddingVertical-rjfia r-textAlign-q4m81j r-width-13qz1uu"
                          dir="auto"
                          disabled={false}
                          enterKeyHint="next"
                          maxLength={1}
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          onKeyDown={[Function]}
                          onSelect={[Function]}
                          placeholder="*"
                          readOnly={false}
                          spellCheck={true}
                          type="tel"
                          value=""
                        />
                      </div>
                      <div
                        className="css-view-1dbjc4n r-alignItems-1awozwy r-display-6koalj r-flex-13awgt0 r-paddingHorizontal-mk0yit"
                      >
                        <input
                          autoCapitalize="sentences"
                          autoComplete="on"
                          autoCorrect="on"
                          className="css-textinput-11aywtz r-borderBottomColor-1uy9gz5 r-borderBottomWidth-wgabs5 r-borderLeftWidth-gxnn5r r-borderRightWidth-fnigne r-borderTopWidth-13yce4e r-fontSize-hiw28u r-marginBottom-p1pxzi r-marginVertical-r0h9e2 r-paddingVertical-rjfia r-textAlign-q4m81j r-width-13qz1uu"
                          dir="auto"
                          disabled={false}
                          enterKeyHint="next"
                          maxLength={1}
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          onKeyDown={[Function]}
                          onSelect={[Function]}
                          placeholder="*"
                          readOnly={false}
                          spellCheck={true}
                          type="tel"
                          value=""
                        />
                      </div>
                      <div
                        className="css-view-1dbjc4n r-alignItems-1awozwy r-display-6koalj r-flex-13awgt0 r-marginLeft-1ozfoo7 r-paddingHorizontal-mk0yit"
                      >
                        <input
                          autoCapitalize="sentences"
                          autoComplete="on"
                          autoCorrect="on"
                          className="css-textinput-11aywtz r-borderBottomColor-1uy9gz5 r-borderBottomWidth-wgabs5 r-borderLeftWidth-gxnn5r r-borderRightWidth-fnigne r-borderTopWidth-13yce4e r-fontSize-hiw28u r-marginBottom-p1pxzi r-marginVertical-r0h9e2 r-paddingVertical-rjfia r-textAlign-q4m81j r-width-13qz1uu"
                          dir="auto"
                          disabled={false}
                          enterKeyHint="next"
                          maxLength={1}
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          onKeyDown={[Function]}
                          onSelect={[Function]}
                          placeholder="*"
                          readOnly={false}
                          spellCheck={true}
                          type="tel"
                          value=""
                        />
                      </div>
                      <div
                        className="css-view-1dbjc4n r-alignItems-1awozwy r-display-6koalj r-flex-13awgt0 r-paddingHorizontal-mk0yit"
                      >
                        <input
                          autoCapitalize="sentences"
                          autoComplete="on"
                          autoCorrect="on"
                          className="css-textinput-11aywtz r-borderBottomColor-1uy9gz5 r-borderBottomWidth-wgabs5 r-borderLeftWidth-gxnn5r r-borderRightWidth-fnigne r-borderTopWidth-13yce4e r-fontSize-hiw28u r-marginBottom-p1pxzi r-marginVertical-r0h9e2 r-paddingVertical-rjfia r-textAlign-q4m81j r-width-13qz1uu"
                          dir="auto"
                          disabled={false}
                          enterKeyHint="next"
                          maxLength={1}
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          onKeyDown={[Function]}
                          onSelect={[Function]}
                          placeholder="*"
                          readOnly={false}
                          spellCheck={true}
                          type="tel"
                          value=""
                        />
                      </div>
                      <div
                        className="css-view-1dbjc4n r-alignItems-1awozwy r-display-6koalj r-flex-13awgt0 r-paddingHorizontal-mk0yit"
                      >
                        <input
                          autoCapitalize="sentences"
                          autoComplete="on"
                          autoCorrect="on"
                          className="css-textinput-11aywtz r-borderBottomColor-1uy9gz5 r-borderBottomWidth-wgabs5 r-borderLeftWidth-gxnn5r r-borderRightWidth-fnigne r-borderTopWidth-13yce4e r-fontSize-hiw28u r-marginBottom-p1pxzi r-marginVertical-r0h9e2 r-paddingVertical-rjfia r-textAlign-q4m81j r-width-13qz1uu"
                          dir="auto"
                          disabled={false}
                          enterKeyHint="next"
                          maxLength={1}
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          onKeyDown={[Function]}
                          onSelect={[Function]}
                          placeholder="*"
                          readOnly={false}
                          spellCheck={true}
                          type="tel"
                          value=""
                        />
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz r-marginVertical-1r5su4o"
                  style={
                    Object {
                      "WebkitBoxPack": "center",
                      "WebkitJustifyContent": "center",
                      "justifyContent": "center",
                      "msFlexPack": "center",
                    }
                  }
                >
                  <div
                    className="css-text-901oao r-color-vaa0vc r-fontFamily-vnw8o6 r-fontSize-1b43r93 r-fontWeight-1it3c9n r-lineHeight-qvk6io r-textAlign-q4m81j r-textDecorationLine-13wfysu r-textTransform-3twk1y"
                    dir="auto"
                  >
                    Please wait 60 seconds until the SMS arrives
                  </div>
                </div>
              </div>
            </div>
            <div
              className="css-view-1dbjc4n r-alignItems-1awozwy r-flexDirection-18u37iz"
              style={
                Object {
                  "WebkitBoxPack": "end",
                  "WebkitJustifyContent": "flex-end",
                  "justifyContent": "flex-end",
                  "msFlexPack": "end",
                }
              }
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="css-view-1dbjc4n r-bottom-1p0dtai r-left-1d2f490 r-pointerEvents-12vffkv r-position-u8s1d r-right-zchlnj r-top-ipm5af"
  />,
]
`;
